ensembleperturbation/perturbation/atcf.py

```diff
@@ -601,56 +727,198 @@ def __init__(self):
+            # bi-section method
+            alpha[Rrat <= 1] *= 0.5 * (1 + (Vr / Vr_test)[Rrat <= 1] ** 2)
+            alpha[Rrat > 1] *= 0.5 * (1 + (Vr_test / Vr)[Rrat > 1] ** 2)
+            if B is not None:
+                # update to and Bg, phi
+                Bg = numpy.log(alpha) / numpy.log(Rrat)
+                phi = 1 + Ro_inv / (Bg * (1 + Ro_inv))
+                phi[phi < 1] = 1
+            else:
+                # update to Rrat
+                Rrat = numpy.exp(numpy.log(alpha) / Bg)
+        if B is not None:
+            return Bg * phi / ((1 + Ro_inv) * numpy.exp(phi - 1))  # B
+        else:
+            return Rmax / Rrat  # isotach_rad
+
+    def get_isotach_properties(self, dataframe: DataFrame) -> Quantity:
+        """ 
+        Return properties of storm and isotach after adjusting for quadrant angle and boundary layer height.
+        LC12: Lin, N., and D. Chavas (2012), On hurricane parametric wind and applications in storm surge modeling, 
+        J. Geophys. Res., 117, D09120, doi:10.1029/2011JD017126 
+        """
+
+        def ts_factor(isotach_rad, Rmax):
+            # factors based on LC12 Figure 2
+            rRm = isotach_rad / Rmax
+            rRm[rRm == 0] = 1e6  # convert a zero value to the upper limit
+            factor = 0.66 + (0.6 - 0.66) * rRm / 2
+            factor[rRm > 2] = 0.6 + (0.55 - 0.6) * (rRm[rRm > 2] - 2) / 4
+            factor[rRm > 6] = 0.55
+            factor[rRm < 1] = 0.3 + (0.66 - 0.3) * rRm[rRm < 1]
+            return factor
+
+        # get Vmax after subtracting speed and adjusting to boundary layer height
+        Vmax, translation_speed = MaximumSustainedWindSpeed().radial_Vmax_at_BL(dataframe)
+        # get Vr for the X-kt isotach in this quadrant
+        Vr = (
+            dataframe['isotach_radius'].values
+            * MaximumSustainedWindSpeed.unit  # LC12 factor and 20-deg CCW rotation
+            - ts_factor(
+                dataframe[self.name].values, dataframe[RadiusOfMaximumWinds.name].values
+            )
+            * translation_speed
+            * numpy.sin(numpy.deg2rad(self.quadrant_angle + 20))
+        ) / BLAdj
+        # get Rmax with units
+        Rmax = dataframe[RadiusOfMaximumWinds.name].values * RadiusOfMaximumWinds.unit
+        # get Coriolis and compute Rossby number (inverse of)
+        f = 2 * omega * numpy.sin(numpy.deg2rad(dataframe['latitude'].values))
+        Ro_inv = (Rmax * f / Vmax).to('dimensionless')
+        # get central pessure and compute standard Holland B parameter as initial guess
+        DelP = dataframe[BackgroundPressure.name] - dataframe[CentralPressure.name]
+        DelP = DelP.values * CentralPressure.unit
+        B = Vmax.to('m/s') ** 2 * AIR_DENSITY * E1 / DelP.to('Pa')
+        # now found the Bg and phi (and subsequent B) that actually fits
+        # through both the isotach and the Rmax value
+        return Vmax, Vr, Rmax, Ro_inv, f, B
+
+    def perturb(
```

Comment:
The perturbation function used to find the new isotach radii in the current quadrant (self) using the GAHM model https://wiki.adcirc.org/Generalized_Asymmetric_Holland_Model.